### PR TITLE
Build Amity Room Allocation system

### PR DESCRIPTION
**Pull request by ivantoz**
#### What does this PR do?
Creates Room allocation system
#### Description of Task to be completed?
Create a command line interface using docopt that does the following using the models.

- [x] create_room <room_name>... 
 -Creates rooms in Amity. Using this command I should be able to create as many rooms as possible by specifying multiple room names after the create_room command.

- [x] add_person <person_name> <FELLOW|STAFF> [wants_accommodation] - Adds a person to the system and allocates the person to a random room. wants_accommodation here is an optional argument which can be either Y or N. The default value if it is not provided is N.
- [x] reallocate_person <person_identifier> <new_room_name> - Reallocate the person with person_identifier to new_room_name.
- [x] load_people  - Adds people to rooms from a txt file. See data.txt for text input format.
- [x] print_allocations [-o=filename] - Prints a list of allocations onto the screen. Specifying the optional -o option here outputs the registered allocations to a txt file. See Appendix 2A for format.
- [x] print_unallocated [-o=filename] - Prints a list of unallocated people to the screen. Specifying the -o option here outputs the information to the txt file provided.
- [x] print_room <room_name> - Prints  the names of all the people in room_name on the screen.
- [x] save_state [--db=sqlite_database] - Persists all the data stored in the app to a SQLite database. Specifying the --db parameter explicitly stores the data in the sqlite_database specified.  
- [x] load_state <sqlite_database> - Loads data from a database into the application.
#### How should this be manually tested?
After cloning the repo and CD into root RUN `pystest tests/test_amity.py`
#### Any background context you want to provide?
Amity has rooms which can be offices or living spaces. An office can accommodate a maximum of 6 people. A living space can accommodate a maximum of 4 people.

A person to be allocated could be a fellow or staff. Staff cannot be allocated living spaces. Fellows have a choice to choose a living space or not.

This system will be used to automatically allocate spaces to people at random.

### CODE REVIEW

File: app/amity.py [hunk 1]
```diff
@@ -15,17 +20,20 @@ def __getitem__(self, key):
 
 
 class Amity(object):
+    # TODO add docstring documentation with examples for doctest
```
Comment (brotich) [inline app/amity.py]: clear TODO items before presenation

File: app/amity.py [hunk 2]
```diff
@@ -34,19 +42,20 @@ def create_room(self, room_type, *args):
             if len(args) >= 1:
```
Comment (brotich) [inline app/amity.py]: this is redundant, list containing more than one item is considered true, so `if args` should be appropriate

File: app/amity.py [hunk 3]
```diff
@@ -34,19 +42,20 @@ def create_room(self, room_type, *args):
...
                 msg = ''
                 for room_name in args:
```
Comment (brotich) [inline app/amity.py]: possible use a more descriptive variable name for `args`

File: app/amity.py [hunk 4]
```diff
@@ -64,141 +73,212 @@ def generate_room(self, r_type):
         available = []
         for room in self.all_rooms:
 
-            if room.room_type.upper() == r_type.upper():
-                if room.room_type.upper() == "LSPACE":
+            if room.room_type == r_type.upper():
```
Comment (brotich) [inline app/amity.py]: possibly use `room_type` instead of `r_type`.

File: app/amity.py [hunk 5]
```diff
@@ -64,141 +73,212 @@ def generate_room(self, r_type):
...
-                staff = Staff(person_name, job_type, wants_accomodation)
+
+                staff = Staff(employee_number.upper(), first_name.upper(), last_name.upper(), job_type.upper(),
+                              wants_accomodation.upper())
                 self.all_people.append(staff)
                 self.staff_list.append(staff)
-                print("{} successfully added to system".format(person_name))
 
                 allocated_office = self.generate_room("OFFICE")
                 if allocated_office is not None:
```
Comment (brotich) [inline app/amity.py]: use `if allocated_office`, None objects are false in python

File: app/amity.py [hunk 6]
```diff
@@ -64,141 +73,212 @@ def generate_room(self, r_type):
...
+
             elif job_type.upper() == "FELLOW":
-                fellow = Fellow(person_name, job_type, wants_accomodation)
+                fellow = Fellow(employee_number.upper(), first_name.upper(), last_name.upper(), job_type.upper(),
+                                wants_accomodation.upper())
                 self.all_people.append(fellow)
                 self.fellows_list.append(fellow)
                 allocated_office = self.generate_room("OFFICE")
                 kudosmsg = ''
                 if allocated_office is not None:
```
Comment (brotich) [inline app/amity.py]: note review above

File: app/amity.py [hunk 7]
```diff
@@ -64,141 +73,212 @@ def generate_room(self, r_type):
...
+                return "Sorry you cannot allocate staff living space!"
+            # check whether the person is already in the allocated room
+            if employee_number.upper() in self.lspace_allocations[new_room.upper()]:
+                return "The Person is already allocated in the requested room"
+            else:
+
+                new_room_occupant_count = [room.occupants for room in self.all_rooms if room.room_name ==
+                                           new_room.upper()]
+
+                if current_room is not None:
```
Comment (brotich) [inline app/amity.py]: No need to check if is not `None`, by default, None is false

File: app/amity.py [hunk 8]
```diff
@@ -64,141 +73,212 @@ def generate_room(self, r_type):
...
+                        self.lspace_allocations[new_room.upper()].append(employee_number.upper())
+                        return "{} has been moved to {}".format(employee_number.upper(), new_room.upper())
+                    else:
+                        return "Sorry the LivingSpace is currently fully occupied!"
+
+        if rtype.upper() == "OFFICE":
+            # check whether the person is already in the allocated room
+            if employee_number.upper() in self.office_allocations[new_room.upper()]:
+                return "The Person is already allocated in the requested room"
+            else:
```
Comment (brotich) [inline app/amity.py]: else statement is redundant, the return statement terminate the function here

File: app/amity.py [hunk 9]
```diff
@@ -207,54 +287,271 @@ def load_people(self, filename):
                         person_data = line.split()
```
Comment (brotich) [inline app/amity.py]: check previous comment on the redundancy  of else statement with `return `

File: app/amity.py [hunk 10]
```diff
@@ -207,54 +287,271 @@ def load_people(self, filename):
...
-        for office, names in self.office_allocations.items():
-            print(office.upper())
-            print("-" * 30)
-            print("\n")
-            for name in names:
-                print(name)
-                print("\n")
+        """Prints a list of allocations onto the screen. Specifying the optional -o option here outputs the registered
+        allocations to a txt file"""
+        if len(self.office_allocations) or len(self.lspace_allocations) != 0:
```
Comment (brotich) [inline app/amity.py]: an empty list by default is falsy ,  so no need to check for length to confirm if its true or false

File: app/amity.py [hunk 11]
```diff
@@ -207,54 +287,271 @@ def load_people(self, filename):
...
+                lastname = [person.last_name for person in self.all_people if person.employee_number.upper() == employee_number]
+                names += firstname[0].capitalize() + " " + lastname[0].capitalize() + ", "
+            output += '\n\t' + bcolors.OKBLUE + names + bcolors.ENDC
+            output += '\n\t' + bcolors.HEADER + bcolors.UNDERLINE + ' ' * 55 + bcolors.ENDC
+            heading2 = 'UNALLOCATED LIVING SPACE'
+            output += '\n\n\t\t\t\t\t' + bcolors.HEADER + bcolors.BOLD + heading2.upper() + '\n' + bcolors.ENDC
+            output += '\t' + bcolors.HEADER + bcolors.UNDERLINE + ' ' * 55 + bcolors.ENDC + '\n'
+            lspace_names = ''
+            for employee_numb in self.lspace_unallocated:
+                fname = [person.first_name for person in self.all_people if
```
Comment (brotich) [inline app/amity.py]: good job on using oneliners

File: app/amity.py [hunk 12]
```diff
@@ -207,54 +287,271 @@ def load_people(self, filename):
...
+                                                person.employee_number.upper() == empno]
+                                    lspace_names += firstname[0].capitalize() + " " + lastname[0].capitalize() + ", "
+
+                                output += '\n\t' + bcolors.OKBLUE + lspace_names + bcolors.ENDC
+                                output += '\n\t' + bcolors.HEADER + bcolors.UNDERLINE + ' ' * 55 + bcolors.ENDC + '\n'
+                            else:
+                                output += "No one is allocated to {} room".format(room_name)
+
+            return output
+        else:
```
Comment (brotich) [inline app/amity.py]: else is redundant here because of the `return` statement

File: tests/test_amity.py [hunk 13]
```diff
@@ -1,146 +1,299 @@
...
-    def test_add_person_not_allow_more_than_capacity_office_allocation(self):
-        people_list = {"Sam": "Fellow", "Gideon": "Fellow", "Charles": "Fellow", "Rogers": "Staff", "Mahad": "Fellow",
-                       "Percila": "Staff", "Brian": "Fellow"}
-        self.amity.create_room("office", "Mordor")
-        for name in people_list:
-            assigning_office = self.amity.add_person(name, people_list[name])
+        self.amity.add_person("cn03", "Ivan", "Kipyegon", "fellow", "y")
+        self.assertEqual(len(self.amity.all_people), person_count_before + 1)
+
+    def test_add_person_duplicate_employee_number(self):
```
Comment (brotich) [inline tests/test_amity.py]: test message should be more descriptive i.e `test_error_adding_person_with_duplicate_employee_number`

File: tests/test_amity.py [hunk 14]
```diff
@@ -1,146 +1,299 @@
...
-        self.amity.create_room("office", "Valhalla")
-        self.amity.reallocate_person("Gideon", "Valhalla")
-        self.assertNotIn("Gideon", self.amity.office_allocations["Oculus"])
-
-    def test_it_prints_unallocated(self):
-        self.amity.print_unallocated('test_print')
-        self.assertTrue(os.path.isfile('test_print.txt'))
-        os.remove('test_print.txt')
+        """Test that a reallocated person is removed from old room"""
+        self.amity.add_person("CN02", "John", "Doe", "fellow", "Y")
```
Comment (brotich) [inline tests/test_amity.py]: Possible redundant tests? this seems like repeated tests for the same feature

File: tests/test_amity.py [hunk 15]
```diff
@@ -1,146 +1,299 @@
...
-        self.assertEqual(self.amity.reallocate_person("Batian7", "Kenya"), "Sorry the office is fully occupied")
+        self.assertEqual(self.amity.reallocate_person("cn05", "Shell"), "Sorry the LivingSpace is currently fully occupied!")
 
     def test_reallocate_staff_to_living_space(self):
-        self.amity.create_room("Office", "Longonot")
-        self.amity.add_person("chelimo", "Staff")
-        self.amity.create_room("lspace", "Shell")
-        allocating_staff_living_space = self.amity.reallocate_person("chelimo", "Shell")
-        self.assertEqual(allocating_staff_living_space, "Sorry you cannot allocate staff living space")
+        """test reallocating staff to a living space room"""
```
Comment (brotich) [inline tests/test_amity.py]: this is a test for living space, the office part seems unnecessary, lets have a discussion on this

File: README.md [hunk 16]
```diff
@@ -0,0 +1,51 @@
...
+Install dependencies
+`$ pip install -r requirements.txt`
+
+**Run the application**
+```
+
+$ python app.py -i
+
+```
+
```
Comment (brotich) [inline README.md:32]: add instructions on how to run the tests

File: app/amity.py [hunk 17]
```diff
@@ -1,34 +1,535 @@
...
+        Adds a person to the system and allocates a random room
+
+        :return:
+
+        """
+
+        all_employees_number = [person.employee_number for person in self.all_people]
+        if employee_number.upper() in all_employees_number:
+            msg = "sorry, this user already exists.please enter valid employee number"
+            print(msg)
```
Comment (brotich) [inline app/amity.py]: check  the print statements

File: app/amity.py [hunk 18]
```diff
@@ -1,34 +1,535 @@
...
+            return "sorry, that room_type does not exist"
+
+    def generate_room(self, room_type):
+        """returns a list of all rooms not full"""
+        available = []
+        for room in self.all_rooms:
+
+            if room.room_type == room_type.upper():
+                if room.room_type == "LSPACE":
+                    if room.occupants < room.max_occupants:
```
Comment (brotich) [inline app/amity.py]: combine the condition statements using and

File: app/amity.py [hunk 19]
```diff
@@ -1,34 +1,535 @@
...
+
+    def load_people(self, filename):
+        """Adds people to rooms from a text file"""
+
+        if os.path.exists(filename):
+            with open(filename) as inputfl:
+
+                content = inputfl.readlines()
+
+                if len(content) == 0:
```
Comment (brotich) [inline app/amity.py]: empty lines is a false condition, try `if not content`

File: app/amity.py [hunk 20]
```diff
@@ -1,34 +1,535 @@
...
+                    return "The file is empty"
+                else:
```
Comment (brotich) [inline app/amity.py]: redundant `else` statement , indent code to match the `if` indentation

File: README.md (hunk 16 above)
Comment (ghost) [reply README.md:32]: you can improve on your README by showing output from running the program such as showing command and output from the program

**Review by ghost — COMMENTED** [review]
Hi Ivan, you've done a good work by covering many edge cases in your application but you need to improve on writing your README to make it easy for someone interact with your program

When running your tests locally using pytest, try using this command:
pytest -v --cov=app --cov-report=xml --cov-report=term-missing

It will help your identify the parts of your programs not covered in your tests helping you refactor the tests and improve coverage. You only need to check the missing lines and create assertions in your test that test against the values already in your code and your coverage will greatly improve way above 80%
